docs: add a guide for using `Component`s (#204)
obsidianmd/obsidian-developer-docs: en/Plugins/Guides/Manage plugin lifecycle.md

## What you'll learn

After completing this guide, you'll be able to:

- Manage the lifecycle of resources in your plugin.
- Use the `Component` class to simplify cleanup and organization.
- Avoid common pitfalls that lead to resource leaks or stale data.
- Understand how components and subcomponents interact in Obsidian.

## Overview

Imagine your plugin needs to respond to window resize events.
A simple (but incorrect implementation!) might look like this:

```ts
// Bad
class MyPlugin extends Plugin {
	onload() {
		window.addEventListener("resize", onResize);
	}
}
```

This works, but it introduces a problem: when the user disabled your plugin, the resize listener remains active! Only when Obsidian is restarted will it be removed. While this might be a fairly innocuous example, in more complex scenarios, this can lead to errors and memory leaks.

A more responsible solution would be to manually remove the listener during `onunload`:

```ts
// Better
class MyPlugin extends Plugin {
	onload() {
		window.addEventListener("resize", onResize);
	}

	onunload() {
		window.removeEventListener("resize", onResize);

> 	}

Reviewer: This example is awkward since onResize is undefined and its very easy to mess up `removeEventListener` (if you don't pass in the exact instance, it won't be properly removed).

}
```

> While this works, it quickly becomes repetitive and error-prone. Fortunately, Obsidian's `Component` system offers a better solution.

Reviewer: This seems a little awkward since the example with onload and onunload is _already_ using functions of Component.

this.registerInterval(id);
```

### `registerDomEvent(el, event, handler)`

Registers a DOM event on an element or window.
Automatically removed on unload.

```ts
this.registerDomEvent(window, "resize", onResize);
```

These methods will ensure that any listener you create is always properly released when your plugin or component unloads.

## The Lifecycle of Objects

A good rule of thumb:
> [!NOTE]
> **Any resource created during `onload` (or later) should be cleaned up in `onunload`.**

This applies to various types of resources, including but not limited to:
- DOM event listeners
- Obsidian event listeners
- Intervals and timeouts
- Workers, network connections, or WASM instances
- Third-party libraries that need explicit disposal

Before adding any resource, you should ask yourself:
> [!NOTE]
> **“Who owns this, and how long should it live?”**

Consider the following scenario, you want to listen for clicks on a button inside a custom view, and keyboard inputs on the entire window.

```ts
class MyView extends ItemView {
	constructor(leaf: WorkspaceLeaf) {
		super(leaf);
		const button = this.containerEl.appendChild(
			createEl("button", { text: "Click me!" })
		);
		// Good

> 		button.addEventListener("click", onButtonClick);

Reviewer 2: The Obsidian native approach for these would be:
```ts
new ButtonComponent(this.containerEl).setButtonText('Click me!').onClick(() => {
	new Notice('Button clicked!');
});

// keyboard events
this.scope = new Scope(this.app.scope); //this does not need to be unloaded, Obsidian takes care of it
this.scope.register(['Ctrl'], 's', () => {
	new Notice('Shortcut pressed!');
})
```

Reviewer: For easy comparisons, the addEventListener example is better. It's more confusing to compare window.addEventListener to Scope. Apples and oranges.

```ts
// Good
class MyPlugin extends Plugin {
	// ...
	onload() {
		this.registerDomEvent(window, "resize", onResize);
	}
}
```

The `registerDomEvent` call automatically removes the listener when your plugin unloads, so you don't have to track it manually. The same applies to other helper methods:

### `registerEvent(eventRef)`

Registers an Obsidian `EventRef` (from `app.vault.on(...)`, etc.).
The event will be automatically detached when unloaded.

```ts
this.registerEvent(this.app.vault.on("modify", onFileModify, this));
```

### `registerInterval(id)`

Registers an interval created by `setInterval`.
All registered intervals are automatically cleared on unload.

```ts
const id = window.setInterval(doSomethingPeriodically, 1000);
this.registerInterval(id);
```

### `registerDomEvent(el, event, handler)`

Registers a DOM event on an element or window.
Automatically removed on unload.

```ts
this.registerDomEvent(window, "resize", onResize);
```

These methods will ensure that any listener you create is always properly released when your plugin or component unloads.

## The Lifecycle of Objects

> 

Reviewer: ```suggestion

```

- Use the `Component` class to simplify cleanup and organization.
- Avoid common pitfalls that cause resource leaks.
- Understand how components and subcomponents interact in Obsidian.

## Overview

Imagine your plugin needs to respond to window resize events.
Here's a simple (but incorrect!) implementation:

```ts
// Bad: the listener is never removed!
class MyPlugin extends Plugin {
	onResize() {
		console.debug("Window resized!");
	}

	onload() {
		window.addEventListener("resize", this.onResize);
	}
}
```

This code works initially, but it introduces a problem: when users disable your plugin, the resize listener stays active! It will continue firing until Obsidian is restarted. In this simple example, the impact might be minor, but in a more complex plugin, this can quickly lead to errors and memory leaks.

A more responsible solution would be to manually remove the listener during `onunload`:

```ts
// Better: but more tedious and error-prone.
class MyPlugin extends Plugin {
	onResize() {
		console.debug("Window resized!");
	}

	onload() {
		window.addEventListener("resize", this.onResize);
	}

	onunload() {
		window.removeEventListener("resize", this.onResize);
	}
}
```

> This works, but as your plugin grows, manually tracking every resource you instantiate becomes unwieldy and error-prone. Obsidian's `Component` system provides a better solution.

Reviewer: Nitpick: you start 2 paragraphs in a row with "This works, but..."

```suggestion
As your plugin grows in complexity, manually tracking every resource you instantiate can become unwieldy and error-prone. Obsidian's `Component` system provides a better solution.
```

## Passing Components to Other APIs

Some APIs in Obsidian require a `Component` parameter to manage the lifecycle of whatever they create.
The most common example is the `MarkdownRenderer.render` method:

```ts
MarkdownRenderer.render(app, "## Some Markdown", containerEl, "", COMPONENT);
```

The `COMPONENT` parameter tells the `MarkdownRenderer` which component 'owns' the rendering context. When the parent component unloads, the renderer cleans up automatically.

### Common Pitfall: Orphaned Components

A frequent mistake is creating a `Component`, solely to pass it to an API, without ever unloading it.

```ts
import { Component, ItemView, MarkdownRenderer } from 'obsidian';

class MyView extends ItemView {
	onload() {		
		// Bad: tempComponent is never unloaded!
		const tempComponent = new Component();
		MarkdownRenderer.render(this.app, "## Some Markdown", this.containerEl, "", tempComponent);
		tempComponent.load();
	}
}
```

Since the `tempComponent` is never unloaded, the MarkdownRenderer will never clean up its resources. It becomes a memory leak.

### Solution: Add or use an existing component

> Instead, you should pass a component that has a clearly defined lifecycle, one that will be loaded and unloaded properly. Since _every_ view is a `Component`, we can simply pass the `View` instance itself via `this`:

Author: Added the suggestion!
